File: bridge/typescript/src/_impl/dir/tree/argument.ts
import type { StaticKey } from "../../../_generated/dir/symbol/key.js";
import type { SymbolKind, SymbolSpace } from "../../../_generated/dir/symbol/symbol.js";
import type { GenericParameter, Parameter } from "../../../_generated/dir/tree/argument.js";
import type { LocalNodeId } from "../../../_generated/dir/tree/node.js";

export const GenericParameterImpl = {
    /** Return the key declared by the parameter. */
    symbolKey(parameter: GenericParameter): StaticKey | undefined {
        // error parameters do not declare names
        if (parameter.kind === "error") {
            return undefined;
        }

        return { kind: "name", name: parameter.name };
    },

    /** Return the symbol space declared by the parameter. */
    symbolSpace(parameter: GenericParameter): SymbolSpace | undefined {
        // type parameters bind in type space
        if (parameter.kind === "type" || parameter.kind === "variadicType") {
            return "type";
        }

        // value parameters bind in value space
        else if (parameter.kind === "value" || parameter.kind === "variadicValue") {
            return "value";
        }
        // error parameters do not bind in a symbol space
        else {
            return undefined;
        }
    },

    /** Return the symbol kind declared by the parameter. */
    symbolKind(parameter: GenericParameter): SymbolKind | undefined {
        // type parameters declare generic type symbols
        if (parameter.kind === "type" || parameter.kind === "variadicType") {
            return "genericTypeParameter";
        }

        // value parameters declare generic value symbols
        else if (parameter.kind === "value" || parameter.kind === "variadicValue") {
            return "genericValueParameter";
        }
        // error parameters do not declare symbols
        else {
            return undefined;
        }
    },
};

export const ParameterImpl = {
    /** Return the key declared by the parameter. */
    symbolKey(parameter: Parameter): StaticKey | undefined {
        // only named parameter forms bind a parameter name
        if (parameter.kind === "named" || parameter.kind === "variadicNamed") {
            return { kind: "name", name: parameter.name };
        }

        return undefined;
    },

    /** Return the parameter default value node. */
    defaultValue(parameter: Parameter): LocalNodeId | undefined {
        // only concrete parameter forms can carry defaults
        if (parameter.kind === "named" || parameter.kind === "pattern") {
            return parameter.default;
        }

        return undefined;
    },
};

File: bridge/python/src/destack/_impl/dir/tree/argument.py
from __future__ import annotations

from destack._impl.model import ModelImpl

from typing import TYPE_CHECKING

# type imports
if TYPE_CHECKING:
    from destack._generated.dir.symbol.key import StaticKey
    from destack._generated.dir.symbol.symbol import SymbolKind, SymbolSpace
    from destack._generated.dir.tree.node import LocalNodeId


class GenericParameterImpl(ModelImpl):
    """Methods for generic parameters."""

    def symbol_key(self) -> StaticKey | None:
        """Return the key declared by the parameter."""

        # error parameters do not declare names
        if self.kind == "error":
            return None

        from destack._generated.dir.symbol.key import StaticKeyName

        return StaticKeyName(name=self.name)

    def symbol_space(self) -> SymbolSpace | None:
        """Return the symbol space declared by the parameter."""

        # type parameters bind in type space
        if self.kind in {"type", "variadicType"}:
            return "type"

        # value parameters bind in value space
        elif self.kind in {"value", "variadicValue"}:
            return "value"
        # error parameters do not bind in a symbol space
        else:
            return None

    def symbol_kind(self) -> SymbolKind | None:
        """Return the symbol kind declared by the parameter."""

        # type parameters declare generic type symbols
        if self.kind in {"type", "variadicType"}:
            return "genericTypeParameter"

        # value parameters declare generic value symbols
        elif self.kind in {"value", "variadicValue"}:
            return "genericValueParameter"
        # error parameters do not declare symbols
        else:
            return None


class ParameterImpl(ModelImpl):
    """Methods for callable parameters."""

    def symbol_key(self) -> StaticKey | None:
        """Return the key declared by the parameter."""

        # only named parameter forms bind a parameter name
        if self.kind in {"named", "variadicNamed"}:
            from destack._generated.dir.symbol.key import StaticKeyName

            return StaticKeyName(name=self.name)

        return None

    def default_value(self) -> LocalNodeId | None:
        """Return the parameter default value node."""

        # only concrete parameter forms can carry defaults
        if self.kind in {"named", "pattern"}:
            return self.default

        return None
